### PR TITLE
middleware2d components rendering in native mode

### PR DESCRIPTION
**Pull request by zhakesi**

### CODE REVIEW

File: native/cocos/core/Root.cpp [hunk 1]
```diff
@@ -302,7 +302,7 @@ bool Root::setRenderPipeline(pipeline::RenderPipeline *rppl /* = nullptr*/) {
     onGlobalPipelineStateChanged();
 
     if (_batcher == nullptr) {
-        _batcher = ccnew Batcher2d(this);
+        _batcher = Batcher2d::getInstance(this);
```
Comment (Zhmz) [inline native/cocos/core/Root.cpp:305]: why use singleton here?
Are there over one batcher2d objects in native before?

File: native/cocos/editor-support/MiddlewareManager.cpp [hunk 2]
```diff
@@ -141,6 +142,18 @@ void MiddlewareManager::render(float dt) {
             buffer->uploadIB();
             buffer->uploadVB();
         }
+
+        uint16_t accID = 65534;
+        Batcher2d *bath2d = Batcher2d::getInstance();
```
Comment (Zhmz) [inline native/cocos/editor-support/MiddlewareManager.cpp:147]: typo

File: native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.cpp [hunk 3]
```diff
@@ -143,43 +146,30 @@ void CCArmatureCacheDisplay::render(float /*dt*/) {
 
     auto *mgr = MiddlewareManager::getInstance();
     if (!mgr->isRendering) return;
+    RenderEntity *entity = (RenderEntity *)_entity;
```
Comment (Zhmz) [inline native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.cpp:149]: how about using static_cast?

File: native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.cpp [hunk 4]
```diff
@@ -229,16 +220,17 @@ void CCArmatureCacheDisplay::render(float /*dt*/) {
...
     handleColor(nowColor);
-
+    int segmentCount = 0;
     for (auto *segment : segments) {
         vertexBytes = segment->vertexFloatCount * sizeof(float);
-
-        // check enough space
-        renderInfo->checkSpace(sizeof(uint32_t) * 6, true);
-
+        curDrawInfo = (RenderDrawInfo *)requestDrawInfo(segmentCount++);
```
Comment (Zhmz) [inline native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.cpp:226]: how about using static_cast?

File: native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.cpp [hunk 5]
```diff
@@ -229,16 +220,17 @@ void CCArmatureCacheDisplay::render(float /*dt*/) {
...
+        entity->addDynamicRenderDrawInfo(curDrawInfo);
         // fill new texture index
-        curTextureIndex = segment->getTexture()->getRealTextureIndex();
-        renderInfo->writeUint32(curTextureIndex);
+        curTexture = (cc::Texture2D*)(segment->getTexture()->getRealTexture());
```
Comment (Zhmz) [inline native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.cpp:229]: how about using static_cast?

File: native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.cpp [hunk 6]
```diff
@@ -260,16 +252,22 @@ void CCArmatureCacheDisplay::render(float /*dt*/) {
                 break;
         }
         // fill new blend src and dst
-        renderInfo->writeUint32(curBlendSrc);
-        renderInfo->writeUint32(curBlendDst);
+        Material *material = (Material *)requestMaterial(curBlendSrc, curBlendDst);
```
Comment (Zhmz) [inline native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.cpp:255]: same as above

File: native/cocos/editor-support/dragonbones-creator-support/CCArmatureDisplay.cpp [hunk 7]
```diff
@@ -101,55 +92,37 @@ void CCArmatureDisplay::dbRender() {
     if (this->_armature->getParent()) {
         return;
     }
+    if (!_entity) {
+        return;
+    }
+    RenderEntity *entity = (RenderEntity *)_entity;
```
Comment (Zhmz) [inline native/cocos/editor-support/dragonbones-creator-support/CCArmatureDisplay.cpp:98]: same as above

File: native/cocos/editor-support/dragonbones/animation/WorldClock.cpp [hunk 8]
```diff
@@ -96,21 +96,21 @@ void WorldClock::render() {
...
-                isOrderDirty = true;
-            } else {
-                maxRenderOrder = renderOrder;
-            }
+            //auto renderOrder = animatable->getRenderOrder();
+            //if (maxRenderOrder > renderOrder) {
+            //    isOrderDirty = true;
+            //} else {
+            //    maxRenderOrder = renderOrder;
+            //}
```
Comment (Zhmz) [inline native/cocos/editor-support/dragonbones/animation/WorldClock.cpp:104]: rebundant code should be removed

File: native/cocos/editor-support/spine-creator-support/SkeletonCacheAnimation.h [hunk 9]
```diff
@@ -122,6 +125,7 @@ class SkeletonCacheAnimation : public cc::RefCounted, public cc::middleware::IMi
     bool _isAniComplete = true;
     std::string _animationName;
     bool _useTint = true;
+    bool _batch = false;
```
Comment (Zhmz) [inline native/cocos/editor-support/spine-creator-support/SkeletonCacheAnimation.h:128]: considering that it is a bool value, it is better to rename it as ‘canBatch’ or 'isBatch'

File: native/cocos/editor-support/MeshBuffer.cpp [hunk 10]
```diff
@@ -24,9 +24,23 @@
...
+    gfx::Attribute{gfx::ATTR_NAME_COLOR, gfx::Format::RGBA8, true},
+};
+
+static const ccstd::vector<gfx::Attribute> ATTRIBUTES_V3F_T2F_C4B_C4B{
+    gfx::Attribute{gfx::ATTR_NAME_POSITION, gfx::Format::RGB32F},
+    gfx::Attribute{gfx::ATTR_NAME_TEX_COORD, gfx::Format::RG32F},
+    gfx::Attribute{gfx::ATTR_NAME_COLOR, gfx::Format::RGBA8, true},
+    gfx::Attribute{gfx::ATTR_NAME_COLOR2, gfx::Format::RGBA8, true},
+};
+
```
Comment (LinYunMo) [inline native/cocos/editor-support/MeshBuffer.cpp:43]: Should we put Attribute  in the same place？

File: cocos/dragon-bones/ArmatureDisplay.ts [hunk 11]
```diff
@@ -1340,7 +1371,7 @@ export class ArmatureDisplay extends UIRenderer {
 
     protected createRenderEntity () {
         const renderEntity = new RenderEntity(RenderEntityType.DYNAMIC);
-        renderEntity.setUseLocal(true);
+        renderEntity.setUseLocal(false);
```
Comment (SantyWang) [inline cocos/dragon-bones/ArmatureDisplay.ts:1374]: Why change use local to false

File: cocos/spine/assembler/simple.ts [hunk 12]
```diff
@@ -684,6 +688,20 @@ function realTimeTraverse (batcher: Batcher2D) {
...
+                _m04 = worldMat.m04;
+                _m12 = worldMat.m12;
+                _m01 = worldMat.m01;
+                _m05 = worldMat.m05;
+                _m13 = worldMat.m13;
+                for (let ii = _vertexFloatOffset, nn = _vertexFloatOffset + _vertexFloatCount; ii < nn; ii += _perVertexSize) {
+                    _x = _vbuf[ii];
+                    _y = _vbuf[ii + 1];
+                    _vbuf[ii] = _x * _m00 + _y * _m04 + _m12;
+                    _vbuf[ii + 1] = _x * _m01 + _y * _m05 + _m13;
```
Comment (SantyWang) [inline cocos/spine/assembler/simple.ts:702]: No need to fill z value ?

File: native/cocos/2d/renderer/Batcher2d.cpp [hunk 13]
```diff
@@ -279,6 +290,7 @@ CC_FORCE_INLINE void Batcher2d::handleIADraw(RenderEntity* entity, RenderDrawInf
 
     UIMeshBuffer* currMeshBuffer = drawInfo->getMeshBuffer();
     currMeshBuffer->setDirty(true);
+    _currMeshBuffer = currMeshBuffer;
```
Comment (SantyWang) [inline native/cocos/2d/renderer/Batcher2d.cpp]: ```handleIADraw``` already has lot of  middle ware specific logic. So I think change name to `handleMiddleWareDraw` will be more clear and lead less misunderstanding

File: native/cocos/editor-support/MeshBuffer.h [hunk 14]
```diff
@@ -84,6 +84,14 @@ class MeshBuffer {
         return _ib;
     }
 
+    void *getUIMeshBuffer() const {
```
Comment (SantyWang) [inline native/cocos/editor-support/MeshBuffer.h:87]: why use void*，it's dangerous

File: native/cocos/editor-support/spine-creator-support/SkeletonCacheAnimation.h [hunk 15]
```diff
@@ -134,5 +138,9 @@ class SkeletonCacheAnimation : public cc::RefCounted, public cc::middleware::IMi
     cc::middleware::IOTypedArray *_sharedBufferOffset = nullptr;
     // Js fill this buffer to send parameter to cpp, avoid to call jsb function.
     cc::middleware::IOTypedArray *_paramsBuffer = nullptr;
+    void *_entity = nullptr;
+    std::vector<void *> _drawInfoArray;
```
Comment (SantyWang) [inline native/cocos/editor-support/spine-creator-support/SkeletonCacheAnimation.h:142]: avoid using ```void*```

File: native/cocos/editor-support/MeshBuffer.cpp (hunk 10 above)
Comment (minggo) [reply native/cocos/editor-support/MeshBuffer.cpp:43]: It is better in define in anonymous namespace.

File: native/cocos/editor-support/MeshBuffer.cpp [hunk 16]
```diff
@@ -86,6 +105,8 @@ void MeshBuffer::uploadIB() {
     auto *rIB = _ibArr[_bufferPos];
     rIB->reset();
     rIB->writeBytes(reinterpret_cast<const char *>(_ib.getBuffer()), _ib.length());
+    UIMeshBuffer *uiMeshBuffer = (UIMeshBuffer *)_uiMeshBufferArr[_bufferPos];
```
Comment (minggo) [inline native/cocos/editor-support/MeshBuffer.cpp:108]: Please use static_cast

File: native/cocos/editor-support/MeshBuffer.cpp [hunk 17]
```diff
@@ -86,6 +105,8 @@ void MeshBuffer::uploadIB() {
...
+    uiMeshBuffer->setIData((uint16_t*)rIB->getBuffer());
```
Comment (minggo) [inline native/cocos/editor-support/MeshBuffer.cpp:109]: The same as above.

File: native/cocos/editor-support/MeshBuffer.cpp [hunk 18]
```diff
@@ -107,4 +132,23 @@ void MeshBuffer::reset() {
...
+        attrs = ATTRIBUTES_V3F_T2F_C4B_C4B;
+    }
+    uiMeshBuffer->initialize(std::move(attrs), true);
+    _uiMeshBufferArr.push_back(uiMeshBuffer);
+}
+
+void MeshBuffer::cleanUIMeshBuffer() {
+    for(auto &buf : _uiMeshBufferArr) {
+        delete (UIMeshBuffer*)buf;
+    }
```
Comment (minggo) [inline native/cocos/editor-support/MeshBuffer.cpp:150]: Can change to
```c++
for(auto *buf : _uiMeshBufferArr) {
         delete buf;
}
```

File: native/cocos/editor-support/MeshBuffer.h [hunk 19]
```diff
@@ -93,10 +101,13 @@ class MeshBuffer {
     void clear();
     void init();
     void afterCleanupHandle();
+    void addUIMeshBuffer();
+    void cleanUIMeshBuffer();
 
     ccstd::vector<IOTypedArray *> _ibArr;
     ccstd::vector<IOTypedArray *> _vbArr;
-    std::size_t _bufferPos = 0;
+    ccstd::vector<void *> _uiMeshBufferArr;
```
Comment (minggo) [inline native/cocos/editor-support/MeshBuffer.h:109]: Why use `void*`?

File: native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.h [hunk 20]
```diff
@@ -91,7 +91,10 @@ class CCArmatureCacheDisplay : public cc::RefCounted, public cc::middleware::IMi
 		 * format |render order|world matrix|
          */
     se_object_ptr getParamsBuffer() const;
-
+    void *requestDrawInfo(int idx);
+    void *requestMaterial(uint16_t blendSrc, uint16_t blendDst);
+    void setMaterial(void *material) { _material = material;};
+    void setRenderEntity(void* entity) { _entity = entity;};
```
Comment (minggo) [inline native/cocos/editor-support/dragonbones-creator-support/CCArmatureCacheDisplay.h:97]: Why use `void*`?

File: native/cocos/editor-support/dragonbones-creator-support/CCArmatureDisplay.cpp [hunk 21]
```diff
@@ -233,8 +206,8 @@ CCArmatureDisplay *CCArmatureDisplay::getRootDisplay() {
 
 void CCArmatureDisplay::traverseArmature(Armature *armature, float parentOpacity) {
     static cc::Mat4 matrixTemp;
+    cc::Mat4 nodeWorldMat = ((RenderEntity *)_entity)->getNode()->getWorldMatrix();
```
Comment (minggo) [inline native/cocos/editor-support/dragonbones-creator-support/CCArmatureDisplay.cpp:209]: Use reference?

File: native/cocos/2d/renderer/Batcher2d.cpp [hunk 22]
```diff
@@ -382,6 +388,46 @@ void Batcher2d::generateBatch(RenderEntity* entity, RenderDrawInfo* drawInfo) {
     _batches.push_back(curdrawBatch);
 }
 
+void Batcher2d::generateBatchForIA(RenderEntity* entity, RenderDrawInfo* drawInfo) {
+    auto layer = entity->getNode()->getLayer();
```
Comment (SantyWang) [inline native/cocos/2d/renderer/Batcher2d.cpp:392]: how different between generateBatch and this ?

**Comment by zhakesi** [comment]
new pull request
* https://github.com/cocos/cocos-engine/pull/12788

**Pull request closed without merging**